Adds support for Let's Encrypt x Certbot

AntociAlin opened the pull request:
# Metadata

- **Fixed Issue**: #51 
- **Contributors**: @AntociAlin 

# Proposed Changes

- Implements the new Let's Encrypt x Certbot module for the new MutableSecurity structure.

# New Functioning

Let's Encrypt x Certbot will be able to generate a certificate for a single domain. It supports only Nginx for the moment.

iosifache commented on mutablesecurity/solutions/implementations/lets_encrypt/code.py:
```diff
@@ -1,31 +1,466 @@
+            return int(output[0]) != 0
+
+
+    IDENTIFIER = "test_request"
```
The ID is vague. Please make it more specific (e.g. `request_via_https`).

AntociAlin replied:
Done as requested.

iosifache reviewed (changes requested):
I checked the functionality and it is mostly correct. In addition to the comments in code, please consider the following:
1. The domain change functionality does not work. If you have the domain `dev.mutablesecurity.io` initially set and then change to `staging.mutablesecurity.io`, the certificate remains the same. It should change as well, with the new domain.
2. Replace all the `/var/log/nginx/https_{UserDomain.get()}_access.log` occurrences with `LogLocation.get()`. This indirection decrease the code size and helps further changes (with the additional level of indirection).
3. If you have pyinfra operations, use them. I saw a lot of `server.shell(command="rm"` despite the fact that you could use `file.file(present=False`.
4. Group the objects by types. See the [file structure in the documentation](https://mutablesecurity.io/docs/developers/module-architecture#file-structure).
5. Check the phrases (descriptions for objects and pyinfra operations) to be valid.
6. Lint the code.
7. Remove the `code.py.old` file.

iosifache commented on mutablesecurity/solutions/implementations/lets_encrypt/code.py:
```diff
@@ -1,31 +1,466 @@
+                commands=["apt-get update && apt-get upgrade"],
+            )
+        except:
+            raise CertbotAlreadyUpdatedException()
```
And one more thing: the try-except mechanism here doesn't help. One pyinfra raises an error in the execution thread, it can be only caught in the main thread (namely when the Leader module executes the operations).

AntociAlin replied:
I am quite a bit lost. How exactly are we supposed to check if the `UPDATE` has been successfully made, if everything should happen only in the Leader module (the main thread)?

iosifache replied:
Check if the version from the `apt` repository is greater than the installed one. If this is not the case, raise the exception from line `465`.

AntociAlin replied:
Oh, thank you.

AntociAlin replied:
Fixed as requested.

AntociAlin commented:
For the first comment, and that is: _he domain change functionality does not work. If you have the domain dev.mutablesecurity.io initially set and then change to staging.mutablesecurity.io, the certificate remains the same. It should change as well, with the new domain_.
Should the change happen exactly at the time the new information has been made into the configuration file? And if yes, what happens to the old certificate? Should it be revoked? Or should we just auto-generate a another one right as the new information is given, without needing to use the `INSTALL` operation?

iosifache commented:
@AntociAlin, as this version of the module supports only one domain, let's just revoke the old one and generate a new certificate.

AntociAlin commented on mutablesecurity/solutions/implementations/lets_encrypt/code.py:
```diff
@@ -1,31 +1,466 @@
+    IDENTIFIER = "email"
+    DESCRIPTION = "The email of the user whom installs Let's Encrypt on the given domain."
+    INFO_TYPE = StringDataType
+    PROPERTIES = [
```
Done.

AntociAlin commented on mutablesecurity/solutions/implementations/lets_encrypt/code.py:
```diff
@@ -1,31 +1,466 @@
+    IDENTIFIER = "domain"
+    DESCRIPTION = "The domain on which the user installs Let's Encrypt."
+    INFO_TYPE = StringDataType
+    PROPERTIES = [
```
Fixed.

AntociAlin commented on mutablesecurity/solutions/implementations/lets_encrypt/code.py:
```diff
@@ -1,31 +1,466 @@
+
+class RevokeReason(BaseInformation):
+    IDENTIFIER = "revoke_reason"
+    DESCRIPTION = "The reason why Let's Encrypt has been removed.\n\nChoose from 'unspecified', 'keycompromise', 'affiliationchanged', 'superseded', 'cessationofoperation', if you wish to change the reason."
```
Fixed.

AntociAlin commented on mutablesecurity/solutions/implementations/lets_encrypt/code.py:
```diff
@@ -1,31 +1,466 @@
+    DESCRIPTION = "Installed version"
+    INFO_TYPE = StringDataType
+    PROPERTIES = [
+        InformationProperties.METRIC,
```
Done.

AntociAlin commented on mutablesecurity/solutions/implementations/lets_encrypt/code.py:
```diff
@@ -1,31 +1,466 @@
+
+
+class FilePresenceRequirement(BaseTest):
+    IDENTIFIER = "presence"
```
Changed to file_presence. Hope it is clear enough.

AntociAlin commented:
I have added the latest commit, where I've fixed almost every issued presented by you. There is one more issue left, and that is the generation and revocation of certificates when the user changes the domain or the email. The issue occurs only at the `SET_INFORMATION` parameter. Furthermore, `INSTALL` and `UNINSTALL` use the same functions and classes as `SET_INFORMATION`, but they work as expected, no issues at all. Please do look into it George-Andrei Iosif
<georgeiosif@live.com>.

AntociAlin commented on mutablesecurity/solutions/implementations/lets_encrypt/code.py:
```diff
@@ -1,31 +1,466 @@
+
+class Version(BaseInformation):
+    class VersionFact(FactBase):
+        # Let's Encrypt returns the major version as an exit code
```
Yes, it was an old comment, I've removed it.

AntociAlin commented:
I've fixed all the issues presented above. Please do look into this PR one more time and if everything is in order, then it should be ready to merge with main.

iosifache reviewed (approved):
All good, we're merging this!
